File: triad-core/src/main/java/com/nhaarman/triad/TriadIntents.java
package com.nhaarman.triad;

import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.Log;
import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

public final class TriadIntents {

    private static final String KEY_SCREEN = "com.nhaarman.triad.screen";
    private static final String KEY_BACKSTACK = "com.nhaarman.triad.backstack";

    private TriadIntents() {
    }

    public static void startWith(
          @NonNull final Intent intent,
          @NonNull final Class<? extends Screen> screen,
          @NonNull final Class<? extends Screen>... screens
    ) {
        if (screens.length == 0) {
            intent.putExtra(KEY_SCREEN, screen.getCanonicalName());
        } else {
            String[] classNames = new String[screens.length + 1];

            classNames[0] = screen.getCanonicalName();
            for (int i = 0; i < screens.length; i++) {
                classNames[i + 1] = screens[i].getCanonicalName();
            }

            intent.putExtra(KEY_BACKSTACK, classNames);
        }
    }

    @Nullable
    public static Screen<?> createScreen(@NonNull final Intent intent) {
        String screenClassName = intent.getStringExtra(KEY_SCREEN);
        if (screenClassName == null) return null;

        return createScreen(intent, screenClassName);
    }

    @NonNull
    private static Screen<?> createScreen(@NonNull final Intent intent, @NonNull final String screenClassName) {
        Class<?> screenClass = createClassForName(screenClassName);
        if (screenClass == null) {
            throw new IllegalStateException("Could not create a class for " + screenClassName);
        }

        Screen<?> screenInstance = createScreenInstanceUsingConstructor(screenClass);
        if (screenInstance != null) return screenInstance;

        screenInstance = createScreenInstanceUsingCreateMethod(screenClass, intent);
        if (screenInstance != null) return screenInstance;

        Class<?> kotlinScreenClass = createClassForName(screenClassName + "Kt");
        if (kotlinScreenClass == null) {
            throw new IllegalStateException("Could not find empty constructor, `create()` or `create(Intent)` method for class " + screenClassName);
        }

        screenInstance = createScreenInstanceUsingCreateMethod(kotlinScreenClass, intent);
        if (screenInstance != null) return screenInstance;

        throw new IllegalStateException("Could not find empty constructor, `create()` or `create(Intent)` method for class " + screenClassName);
    }

    @Nullable
    public static Backstack createBackstack(@NonNull final Intent intent) {
        String[] backstack = intent.getStringArrayExtra(KEY_BACKSTACK);
        if (backstack == null) return null;

        Screen<?>[] screens = new Screen[backstack.length];
        for (int i = 0; i < backstack.length; i++) {
            screens[i] = createScreen(intent, backstack[i]);
        }

        return Backstack.of(screens);
    }

    @Nullable
    private static Class<?> createClassForName(@NonNull final String className) {
        try {
            return Class.forName(className);
        } catch (ClassNotFoundException e) {
            Log.w("Triad", "Could not create class for name " + className, e);
            return null;
        }
    }

    @Nullable
    private static Screen<?> createScreenInstanceUsingConstructor(@NonNull final Class<?> screenClass) {
        Constructor<?> constructor = parameterlessConstructor(screenClass);
        if (constructor == null) return null;

        try {
            return (Screen<?>) screenClass.newInstance();
        } catch (InstantiationException e) {
            Log.w("Triad", "Could not create screen instance for " + screenClass.getCanonicalName(), e);
            return null;
        } catch (IllegalAccessException e) {
            Log.w("Triad", "Could not create screen instance for " + screenClass.getCanonicalName(), e);
            return null;
        }
    }

    @Nullable
    private static Constructor<?> parameterlessConstructor(@NonNull final Class<?> screenClass) {
        try {
            return screenClass.getConstructor();
        } catch (NoSuchMethodException e) {
            return null;
        }
    }

    @Nullable
    private static Screen<?> createScreenInstanceUsingCreateMethod(@NonNull final Class<?> screenClass, @NonNull final Intent intent) {
        Method createMethod = null;
        for (final Method method : screenClass.getMethods()) {
            if (method.getName().equals("create")) {
                createMethod = method;
            }
        }

        if (createMethod == null) return null;

        try {
            if (createMethod.getParameterTypes().length == 1) {
                return (Screen<?>) createMethod.invoke(null, intent);
            } else {
                return (Screen<?>) createMethod.invoke(null);
            }
        } catch (InvocationTargetException e) {
            Log.w("Triad", "Could not create screen instance for " + screenClass.getCanonicalName(), e);
            return null;
        } catch (IllegalAccessException e) {
            Log.w("Triad", "Could not create screen instance for " + screenClass.getCanonicalName(), e);
            return null;
        }
    }
}

File: triad-core/src/main/kotlin/com/nhaarman/triad/TriadIntents.kt
package com.nhaarman.triad

import android.content.Intent

/**
 * Enriches this [Intent] with the necessary information to start given [Screen]s.
 * Whenever a new [android.app.Activity] is started with this [Intent], the
 * Triad backstack is populated with these [Screen]s.
 *
 * To be able to instantiate a [Screen], it must have one of the following:
 *
 *  - Either a parameterless constructor;
 *  - A static function [create(Intent)] in its companion object;
 *  - Or a top level function [create(Intent)] in the same file.
 */
fun Intent.startWith(vararg screens: Class<out Screen<*>>) {
    if (screens.size == 1) {
        putExtra("com.nhaarman.triad.screen", screens[0].canonicalName)
    } else {
        putExtra("com.nhaarman.triad.backstack", screens.map { it.canonicalName }.toTypedArray())
    }
}

/**
 * Creates a function that is able to create a single [Screen],
 * as set up by [startWith]. Returns [null] if no screen data
 * can be found.
 */
internal fun Intent.createScreen(): (() -> Screen<*>)? =
      getStringExtra("com.nhaarman.triad.screen")?.createMethod(this)

private fun Class<*>.parameterlessConstructor(): (() -> Screen<*>)? {
    return try {
        getConstructor()?.let {
            { it.newInstance() as Screen<*> }
        }
    } catch(e: NoSuchMethodException) {
        null
    }
}

private fun String.createMethod(intent: Intent): () -> Screen<*> =
      createClass()?.parameterlessConstructor() ?:
            createClass()?.createMethod(intent) ?:
            "${this}Kt".createClass()?.createMethod(intent) ?:
            throw error("Could not find empty constructor or `create(Intent)` method for class $this.")

private fun String.createClass(): Class<*>? {
    return try {
        Class.forName(this)
    } catch (e: ClassNotFoundException) {
        null
    }
}

internal fun Intent.createBackstack(): (() -> Backstack)? {
    return getStringArrayExtra("com.nhaarman.triad.backstack")?.let {
        val creates = it.map { it.createMethod(this) }
        return {
            Backstack.of(*creates.map { it() }.toTypedArray())
        }
    }
}

private fun Class<*>.createMethod(intent: Intent): (() -> Screen<*>)? {
    return methods.filter { it.name == "create" }.firstOrNull().let {
        if (it != null) {
            { (if (it.parameterTypes.size == 1) it(null, intent) else it(null)) as Screen <*> }
        } else {
            null
        }
    }
}

